eclipse-collections/src/main/java/org/eclipse/collections/impl/collector/Collectors2.java

```diff
@@ -784,6 +784,40 @@ private Collectors2()
                 EMPTY_CHARACTERISTICS);
     }
 
+    /**
+     * <p>Same as {@link #groupBy(Function, Supplier)}, except the result of evaluating groupBy function should return a
+     * unique key, or else an exception is thrown.</p>
+     *
+     * <p>Equivalent to using @{@link RichIterable#groupByUniqueKey(Function, MutableMap)}</p>
+     */
+    public static <T, K, R extends MutableMultimap<K, T>> Collector<T, ?, R> groupByUniqueKey(
+            Function<? super T, ? extends K> groupBy,
+            Supplier<R> supplier)
+    {
+        return Collector.of(
+                supplier,
+                (map, each) ->
+                {
+                    K key = groupBy.valueOf(each);
+                    if (map.containsKey(key))
+                    {
+                        throw new IllegalStateException("Key " + key + " already exists in map!");
+                    }
+                    map.put(key, each);
+                },
+                (r1, r2) ->
+                {
+                    K nonUniqueKey = r2.keysView().detect(key -> r2.get(key).size() > 1);
```

Comment:
You could use `putAll` and check the size of the result and check that it has increased by the size of the Map being added.  But this will not help you figure out which key was a duplicate.  I would just use `r2.forEachKeyValue((k, v) -> code to put and check for dupes)`.